\documentclass[11pt]{article}
\renewcommand{\arraystretch}{1.5} % Default value: 1
\usepackage{sectsty}
\allsectionsfont{\color{blue}\fontfamily{lmss}\selectfont}
\usepackage{fontspec}
\setmainfont{XCharter}

\usepackage{listings}

\lstset{
basicstyle=\small\ttfamily,
tabsize=8,
columns=flexible,
breaklines=true,
frame=tb,
rulecolor=\color[rgb]{0.8,0.8,0.7},
backgroundcolor=\color[rgb]{1,1,0.91},
postbreak=\raisebox{0ex}[0ex][0ex]{\ensuremath{\color{red}\hookrightarrow\space}}
}
\usepackage{fontawesome}


\usepackage{mdframed}
\newmdenv[
  backgroundcolor=gray,
  fontcolor=white,
  nobreak=true,
]{terminalinput}



\usepackage{parskip}


    \usepackage[breakable]{tcolorbox}
    \usepackage{parskip} % Stop auto-indenting (to mimic markdown behaviour)

    \usepackage{iftex}
    \ifPDFTeX
    	\usepackage[T1]{fontenc}
    	\usepackage{mathpazo}
    \else
    	\usepackage{fontspec}
    \fi

    % Basic figure setup, for now with no caption control since it's done
    % automatically by Pandoc (which extracts ![](path) syntax from Markdown).
    \usepackage{graphicx}
    % Maintain compatibility with old templates. Remove in nbconvert 6.0
    \let\Oldincludegraphics\includegraphics
    % Ensure that by default, figures have no caption (until we provide a
    % proper Figure object with a Caption API and a way to capture that
    % in the conversion process - todo).
    \usepackage{caption}
    \DeclareCaptionFormat{nocaption}{}
    \captionsetup{labelformat=nolabel, textfont=bf}

    \usepackage{float}
    \floatplacement{figure}{H} % forces figures to be placed at the correct location
    \usepackage{xcolor} % Allow colors to be defined
    \usepackage{enumerate} % Needed for markdown enumerations to work
    \usepackage{geometry} % Used to adjust the document margins
    \usepackage{amsmath} % Equations
    \usepackage{amssymb} % Equations
    \usepackage{textcomp} % defines textquotesingle
    % Hack from http://tex.stackexchange.com/a/47451/13684:
    \AtBeginDocument{%
        \def\PYZsq{\textquotesingle}% Upright quotes in Pygmentized code
    }
    \usepackage{upquote} % Upright quotes for verbatim code
    \usepackage{eurosym} % defines \euro
    \usepackage[mathletters]{ucs} % Extended unicode (utf-8) support
    \usepackage{fancyvrb} % verbatim replacement that allows latex
    \usepackage{grffile} % extends the file name processing of package graphics
                         % to support a larger range
    \makeatletter % fix for old versions of grffile with XeLaTeX
    \@ifpackagelater{grffile}{2019/11/01}
    {
      % Do nothing on new versions
    }
    {
      \def\Gread@@xetex#1{%
        \IfFileExists{"\Gin@base".bb}%
        {\Gread@eps{\Gin@base.bb}}%
        {\Gread@@xetex@aux#1}%
      }
    }
    \makeatother
    \usepackage[Export]{adjustbox} % Used to constrain images to a maximum size
    \adjustboxset{max size={0.9\linewidth}{0.9\paperheight}}

    % The hyperref package gives us a pdf with properly built
    % internal navigation ('pdf bookmarks' for the table of contents,
    % internal cross-reference links, web links for URLs, etc.)
    \usepackage{hyperref}
    % The default LaTeX title has an obnoxious amount of whitespace. By default,
    % titling removes some of it. It also provides customization options.
    \usepackage{titling}
    \usepackage{longtable} % longtable support required by pandoc >1.10
    \usepackage{booktabs}  % table support for pandoc > 1.12.2
    \usepackage[inline]{enumitem} % IRkernel/repr support (it uses the enumerate* environment)
    \usepackage[normalem]{ulem} % ulem is needed to support strikethroughs (\sout)
                                % normalem makes italics be italics, not underlines
    \usepackage{mathrsfs}



    % Colors for the hyperref package
    \definecolor{urlcolor}{rgb}{0,.145,.698}
    \definecolor{linkcolor}{rgb}{.71,0.21,0.01}
    \definecolor{citecolor}{rgb}{.12,.54,.11}

    % ANSI colors
    \definecolor{ansi-black}{HTML}{3E424D}
    \definecolor{ansi-black-intense}{HTML}{282C36}
    \definecolor{ansi-red}{HTML}{E75C58}
    \definecolor{ansi-red-intense}{HTML}{B22B31}
    \definecolor{ansi-green}{HTML}{00A250}
    \definecolor{ansi-green-intense}{HTML}{007427}
    \definecolor{ansi-yellow}{HTML}{DDB62B}
    \definecolor{ansi-yellow-intense}{HTML}{B27D12}
    \definecolor{ansi-blue}{HTML}{208FFB}
    \definecolor{ansi-blue-intense}{HTML}{0065CA}
    \definecolor{ansi-magenta}{HTML}{D160C4}
    \definecolor{ansi-magenta-intense}{HTML}{A03196}
    \definecolor{ansi-cyan}{HTML}{60C6C8}
    \definecolor{ansi-cyan-intense}{HTML}{258F8F}
    \definecolor{ansi-white}{HTML}{C5C1B4}
    \definecolor{ansi-white-intense}{HTML}{A1A6B2}
    \definecolor{ansi-default-inverse-fg}{HTML}{FFFFFF}
    \definecolor{ansi-default-inverse-bg}{HTML}{000000}

    % common color for the border for error outputs.
    \definecolor{outerrorbackground}{HTML}{FFDFDF}

    % commands and environments needed by pandoc snippets
    % extracted from the output of `pandoc -s`
    \providecommand{\tightlist}{%
      \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
    \DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
    % Add ',fontsize=\small' for more characters per line
    \newenvironment{Shaded}{}{}
    \newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{\textbf{{#1}}}}
    \newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.56,0.13,0.00}{{#1}}}
    \newcommand{\DecValTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{{#1}}}
    \newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{{#1}}}
    \newcommand{\FloatTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{{#1}}}
    \newcommand{\CharTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\StringTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\CommentTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textit{{#1}}}}
    \newcommand{\OtherTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{{#1}}}
    \newcommand{\AlertTok}[1]{\textcolor[rgb]{1.00,0.00,0.00}{\textbf{{#1}}}}
    \newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.02,0.16,0.49}{{#1}}}
    \newcommand{\RegionMarkerTok}[1]{{#1}}
    \newcommand{\ErrorTok}[1]{\textcolor[rgb]{1.00,0.00,0.00}{\textbf{{#1}}}}
    \newcommand{\NormalTok}[1]{{#1}}

    % Additional commands for more recent versions of Pandoc
    \newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.53,0.00,0.00}{{#1}}}
    \newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.73,0.40,0.53}{{#1}}}
    \newcommand{\ImportTok}[1]{{#1}}
    \newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.73,0.13,0.13}{\textit{{#1}}}}
    \newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    \newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    \newcommand{\VariableTok}[1]{\textcolor[rgb]{0.10,0.09,0.49}{{#1}}}
    \newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{\textbf{{#1}}}}
    \newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.40,0.40,0.40}{{#1}}}
    \newcommand{\BuiltInTok}[1]{{#1}}
    \newcommand{\ExtensionTok}[1]{{#1}}
    \newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.74,0.48,0.00}{{#1}}}
    \newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.49,0.56,0.16}{{#1}}}
    \newcommand{\InformationTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    \newcommand{\WarningTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}


    % Define a nice break command that doesn't care if a line doesn't already
    % exist.
    \def\br{\hspace*{\fill} \\* }
    % Math Jax compatibility definitions
    \def\gt{>}
    \def\lt{<}
    \let\Oldtex\TeX
    \let\Oldlatex\LaTeX
    \renewcommand{\TeX}{\textrm{\Oldtex}}
    \renewcommand{\LaTeX}{\textrm{\Oldlatex}}
    % Document parameters
    % Document title
    \title{index}





% Pygments definitions
\makeatletter
\def\PY@reset{\let\PY@it=\relax \let\PY@bf=\relax%
    \let\PY@ul=\relax \let\PY@tc=\relax%
    \let\PY@bc=\relax \let\PY@ff=\relax}
\def\PY@tok#1{\csname PY@tok@#1\endcsname}
\def\PY@toks#1+{\ifx\relax#1\empty\else%
    \PY@tok{#1}\expandafter\PY@toks\fi}
\def\PY@do#1{\PY@bc{\PY@tc{\PY@ul{%
    \PY@it{\PY@bf{\PY@ff{#1}}}}}}}
\def\PY#1#2{\PY@reset\PY@toks#1+\relax+\PY@do{#2}}

\expandafter\def\csname PY@tok@w\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.73,0.73}{##1}}}
\expandafter\def\csname PY@tok@c\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@cp\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.74,0.48,0.00}{##1}}}
\expandafter\def\csname PY@tok@k\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kp\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kt\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.69,0.00,0.25}{##1}}}
\expandafter\def\csname PY@tok@o\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@ow\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.67,0.13,1.00}{##1}}}
\expandafter\def\csname PY@tok@nb\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@nf\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@nc\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@nn\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@ne\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.82,0.25,0.23}{##1}}}
\expandafter\def\csname PY@tok@nv\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@no\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.53,0.00,0.00}{##1}}}
\expandafter\def\csname PY@tok@nl\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.63,0.63,0.00}{##1}}}
\expandafter\def\csname PY@tok@ni\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.60,0.60,0.60}{##1}}}
\expandafter\def\csname PY@tok@na\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.49,0.56,0.16}{##1}}}
\expandafter\def\csname PY@tok@nt\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@nd\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.67,0.13,1.00}{##1}}}
\expandafter\def\csname PY@tok@s\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sd\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@si\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.73,0.40,0.53}{##1}}}
\expandafter\def\csname PY@tok@se\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.73,0.40,0.13}{##1}}}
\expandafter\def\csname PY@tok@sr\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.40,0.53}{##1}}}
\expandafter\def\csname PY@tok@ss\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@sx\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@m\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@gh\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,0.50}{##1}}}
\expandafter\def\csname PY@tok@gu\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.50,0.00,0.50}{##1}}}
\expandafter\def\csname PY@tok@gd\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.63,0.00,0.00}{##1}}}
\expandafter\def\csname PY@tok@gi\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.63,0.00}{##1}}}
\expandafter\def\csname PY@tok@gr\endcsname{\def\PY@tc##1{\textcolor[rgb]{1.00,0.00,0.00}{##1}}}
\expandafter\def\csname PY@tok@ge\endcsname{\let\PY@it=\textit}
\expandafter\def\csname PY@tok@gs\endcsname{\let\PY@bf=\textbf}
\expandafter\def\csname PY@tok@gp\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,0.50}{##1}}}
\expandafter\def\csname PY@tok@go\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.53,0.53,0.53}{##1}}}
\expandafter\def\csname PY@tok@gt\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.27,0.87}{##1}}}
\expandafter\def\csname PY@tok@err\endcsname{\def\PY@bc##1{\setlength{\fboxsep}{0pt}\fcolorbox[rgb]{1.00,0.00,0.00}{1,1,1}{\strut ##1}}}
\expandafter\def\csname PY@tok@kc\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kd\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kn\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kr\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@bp\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@fm\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@vc\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@vg\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@vi\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@vm\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@sa\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sb\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sc\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@dl\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@s2\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sh\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@s1\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@mb\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@mf\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@mh\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@mi\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@il\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@mo\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@ch\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@cm\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@cpf\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@c1\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@cs\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}

\def\PYZbs{\char`\\}
\def\PYZus{\char`\_}
\def\PYZob{\char`\{}
\def\PYZcb{\char`\}}
\def\PYZca{\char`\^}
\def\PYZam{\char`\&}
\def\PYZlt{\char`\<}
\def\PYZgt{\char`\>}
\def\PYZsh{\char`\#}
\def\PYZpc{\char`\%}
\def\PYZdl{\char`\$}
\def\PYZhy{\char`\-}
\def\PYZsq{\char`\'}
\def\PYZdq{\char`\"}
\def\PYZti{\char`\~}
% for compatibility with earlier versions
\def\PYZat{@}
\def\PYZlb{[}
\def\PYZrb{]}
\makeatother


    % For linebreaks inside Verbatim environment from package fancyvrb.
    \makeatletter
        \newbox\Wrappedcontinuationbox
        \newbox\Wrappedvisiblespacebox
        \newcommand*\Wrappedvisiblespace {\textcolor{red}{\textvisiblespace}}
        \newcommand*\Wrappedcontinuationsymbol {\textcolor{red}{\llap{\tiny$\m@th\hookrightarrow$}}}
        \newcommand*\Wrappedcontinuationindent {3ex }
        \newcommand*\Wrappedafterbreak {\kern\Wrappedcontinuationindent\copy\Wrappedcontinuationbox}
        % Take advantage of the already applied Pygments mark-up to insert
        % potential linebreaks for TeX processing.
        %        {, <, #, %, $, ' and ": go to next line.
        %        _, }, ^, &, >, - and ~: stay at end of broken line.
        % Use of \textquotesingle for straight quote.
        \newcommand*\Wrappedbreaksatspecials {%
            \def\PYGZus{\discretionary{\char`\_}{\Wrappedafterbreak}{\char`\_}}%
            \def\PYGZob{\discretionary{}{\Wrappedafterbreak\char`\{}{\char`\{}}%
            \def\PYGZcb{\discretionary{\char`\}}{\Wrappedafterbreak}{\char`\}}}%
            \def\PYGZca{\discretionary{\char`\^}{\Wrappedafterbreak}{\char`\^}}%
            \def\PYGZam{\discretionary{\char`\&}{\Wrappedafterbreak}{\char`\&}}%
            \def\PYGZlt{\discretionary{}{\Wrappedafterbreak\char`\<}{\char`\<}}%
            \def\PYGZgt{\discretionary{\char`\>}{\Wrappedafterbreak}{\char`\>}}%
            \def\PYGZsh{\discretionary{}{\Wrappedafterbreak\char`\#}{\char`\#}}%
            \def\PYGZpc{\discretionary{}{\Wrappedafterbreak\char`\%}{\char`\%}}%
            \def\PYGZdl{\discretionary{}{\Wrappedafterbreak\char`\$}{\char`\$}}%
            \def\PYGZhy{\discretionary{\char`\-}{\Wrappedafterbreak}{\char`\-}}%
            \def\PYGZsq{\discretionary{}{\Wrappedafterbreak\textquotesingle}{\textquotesingle}}%
            \def\PYGZdq{\discretionary{}{\Wrappedafterbreak\char`\"}{\char`\"}}%
            \def\PYGZti{\discretionary{\char`\~}{\Wrappedafterbreak}{\char`\~}}%
        }
        % Some characters . , ; ? ! / are not pygmentized.
        % This macro makes them "active" and they will insert potential linebreaks
        \newcommand*\Wrappedbreaksatpunct {%
            \lccode`\~`\.\lowercase{\def~}{\discretionary{\hbox{\char`\.}}{\Wrappedafterbreak}{\hbox{\char`\.}}}%
            \lccode`\~`\,\lowercase{\def~}{\discretionary{\hbox{\char`\,}}{\Wrappedafterbreak}{\hbox{\char`\,}}}%
            \lccode`\~`\;\lowercase{\def~}{\discretionary{\hbox{\char`\;}}{\Wrappedafterbreak}{\hbox{\char`\;}}}%
            \lccode`\~`\:\lowercase{\def~}{\discretionary{\hbox{\char`\:}}{\Wrappedafterbreak}{\hbox{\char`\:}}}%
            \lccode`\~`\?\lowercase{\def~}{\discretionary{\hbox{\char`\?}}{\Wrappedafterbreak}{\hbox{\char`\?}}}%
            \lccode`\~`\!\lowercase{\def~}{\discretionary{\hbox{\char`\!}}{\Wrappedafterbreak}{\hbox{\char`\!}}}%
            \lccode`\~`\/\lowercase{\def~}{\discretionary{\hbox{\char`\/}}{\Wrappedafterbreak}{\hbox{\char`\/}}}%
            \catcode`\.\active
            \catcode`\,\active
            \catcode`\;\active
            \catcode`\:\active
            \catcode`\?\active
            \catcode`\!\active
            \catcode`\/\active
            \lccode`\~`\~
        }
    \makeatother

    \let\OriginalVerbatim=\Verbatim
    \makeatletter
    \renewcommand{\Verbatim}[1][1]{%
        %\parskip\z@skip
        \sbox\Wrappedcontinuationbox {\Wrappedcontinuationsymbol}%
        \sbox\Wrappedvisiblespacebox {\FV@SetupFont\Wrappedvisiblespace}%
        \def\FancyVerbFormatLine ##1{\hsize\linewidth
            \vtop{\raggedright\hyphenpenalty\z@\exhyphenpenalty\z@
                \doublehyphendemerits\z@\finalhyphendemerits\z@
                \strut ##1\strut}%
        }%
        % If the linebreak is at a space, the latter will be displayed as visible
        % space at end of first line, and a continuation symbol starts next line.
        % Stretch/shrink are however usually zero for typewriter font.
        \def\FV@Space {%
            \nobreak\hskip\z@ plus\fontdimen3\font minus\fontdimen4\font
            \discretionary{\copy\Wrappedvisiblespacebox}{\Wrappedafterbreak}
            {\kern\fontdimen2\font}%
        }%

        % Allow breaks at special characters using \PYG... macros.
        \Wrappedbreaksatspecials
        % Breaks at punctuation characters . , ; ? ! and / need catcode=\active
        \OriginalVerbatim[#1,codes*=\Wrappedbreaksatpunct]%
    }
    \makeatother

    % Exact colors from NB
    \definecolor{incolor}{HTML}{303F9F}
    \definecolor{outcolor}{HTML}{D84315}
    \definecolor{cellborder}{HTML}{CFCFCF}
    \definecolor{cellbackground}{HTML}{F7F7F7}

    % prompt
    \makeatletter
    \newcommand{\boxspacing}{\kern\kvtcb@left@rule\kern\kvtcb@boxsep}
    \makeatother
    \newcommand{\prompt}[4]{
        {\ttfamily\llap{{\color{blue}\LARGE\faKeyboardO\hspace{3pt}#4}}\vspace{-\baselineskip}}
    }



    % Prevent overflowing lines due to hard-to-break entities
    \sloppy
    % Setup hyperref package
    \hypersetup{
      breaklinks=true,  % so long urls are correctly broken across lines
      colorlinks=true,
      urlcolor=urlcolor,
      linkcolor=linkcolor,
      citecolor=citecolor,
      }
    % Slightly bigger margins than the latex defaults

    \geometry{verbose,tmargin=1in,bmargin=1in,lmargin=1in,rmargin=1in}



\renewcommand{\PY}[2]{{#2}}
\usepackage{fancyhdr}
\pagestyle{fancy}
\rhead{\color{gray}\sf\small\rightmark}
\lhead{\nouppercase{\color{gray}\sf\small\leftmark}}
\cfoot{\color{gray}\sf\thepage}
\renewcommand{\footrulewidth}{1pt}
\begin{document}





    \hypertarget{structural-variation-calling}{%
\section{Structural Variation
Calling}\label{structural-variation-calling}}

\hypertarget{introduction}{%
\subsection{Introduction}\label{introduction}}

Structural variants (SVs) are large genomic alterations of at least 50
bp or larger in size. There are several types of SVs, including
deletions, duplications, insertions, inversions, and translocations
which describe different combinations of DNA gains, losses, or
rearrangements. Copy number variations (CNVs) are a particular subtype
of SV mainly represented by deletions and duplications. SVs are
typically described as single events, although more complex scenarios
involving combinations of SV types exist.

    \hypertarget{learning-outcomes}{%
\subsection{Learning outcomes}\label{learning-outcomes}}

On completion of the tutorial, you can expect to be able to:

\begin{itemize}
\tightlist
\item
  Call structural variants using standard tools
\item
  Visualise structural variants using standard tools
\item
  Call structural variants from long read data
\item
  Use bedtools to do regional comparisons over genomic co-ordinates
\end{itemize}

\hypertarget{tutorial-sections}{%
\subsection{Tutorial sections}\label{tutorial-sections}}

This tutorial comprises the following sections: 1.
\href{vcfs.ipynb}{Looking at structural variants in VCF} 2.
\href{sv-calling.ipynb}{Calling structural variants}\\
3. \href{long-reads.ipynb}{Structural variants from long reads}\\
4. \href{bedtools.ipynb}{Bedtools}

\hypertarget{authors}{%
\subsection{Authors}\label{authors}}

This tutorial was written by
\href{https://github.com/jacquikeane}{Jacqui Keane} based on material
from \href{https://github.com/tk2}{Thomas Keane}.

\hypertarget{running-the-commands-in-this-tutorial}{%
\subsection{Running the commands in this
tutorial}\label{running-the-commands-in-this-tutorial}}

You can follow this tutorial by typing all the commands you see into a
terminal window. This is similar to the ``Command Prompt'' window on MS
Windows systems, which allows the user to type DOS commands to manage
files.

To get started, open a new terminal window and type the command below:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{1}{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{n+nb}{cd} /home/manager/course\PYZus{}data/structural\PYZus{}variation/data
\end{Verbatim}
\end{tcolorbox}

    \hypertarget{lets-get-started}{%
\subsection{Let's get started!}\label{lets-get-started}}

This tutorial requires that you have breakdancer, lumpy, minimap2,
sniffles, bedtools and igv installed on your computer. These are already
installed on the virtual machine you are using. To check that these are
installed, run the following commands:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{2}{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
breakdancer\PYZhy{}max \PYZhy{}h
\end{Verbatim}
\end{tcolorbox}

    \begin{Verbatim}[commandchars=\\\{\}, frame=single, framerule=2mm, rulecolor=\color{outerrorbackground}]
    \end{Verbatim}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
lumpy \PYZhy{}\PYZhy{}help
\end{Verbatim}
\end{tcolorbox}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
minimap2 \PYZhy{}\PYZhy{}help
\end{Verbatim}
\end{tcolorbox}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
sniffles \PYZhy{}\PYZhy{}help
\end{Verbatim}
\end{tcolorbox}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
bedtools \PYZhy{}\PYZhy{}help
\end{Verbatim}
\end{tcolorbox}

    This should return the help message for software breakdancer, lumpy,
minimap2, sniffles and bedtools respectively.

If after this course you would like to download and install this
software the instructions can be found at the links below, alternatively
we recommend \href{https://bioconda.github.io/}{bioconda} for the
installation and management of your bioinformatics software.

\begin{itemize}
\tightlist
\item
  The \href{http://breakdancer.sourceforge.net/}{breakdancer website}
\item
  The \href{https://github.com/arq5x/lumpy-sv}{lumpy github page}
\item
  The \href{https://lh3.github.io/minimap2/}{minimap2 website}
\item
  The \href{https://github.com/fritzsedlazeck/Sniffles/wiki}{sniffles
  website}
\item
  The \href{https://bedtools.readthedocs.io/}{bedtools website}
\end{itemize}

To get started with the tutorial, go to the first section:
\href{vcfs.ipynb}{Looking at structural variants in VCF}


    % Add a bibliography block to the postdoc



\newpage





    \hypertarget{looking-at-structural-variants-in-vcf}{%
\section{Looking at Structural Variants in
VCF}\label{looking-at-structural-variants-in-vcf}}

Structural variants can be stored in VCF files. In this part of the
tutorial, we will look at how these are represented in a VCF file.

First, check you are in the correct directory:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{1}{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{n+nb}{pwd}
\end{Verbatim}
\end{tcolorbox}

    It should display something like:

\texttt{/home/manager/course\_data/structural\_variation/data}

    Navigate to the \texttt{exercise1} directory:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{n+nb}{cd} exercise1
\end{Verbatim}
\end{tcolorbox}

    There is a VCF file called \texttt{ERR1015121.vcf} that was produced
using the Lumpy SV calling software. Look at the VCF file using the
\texttt{less} command and answer the questions that follow:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
less ERR1015121.vcf
\end{Verbatim}
\end{tcolorbox}

    \hypertarget{excercises}{%
\subsection{Excercises}\label{excercises}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  What does the CIPOS format tag indicate?
\item
  What does the PE tag indicate?
\item
  What tag is used to describe an inversion event?
\item
  What tag is used to describe a duplication event?
\item
  How many deletions were called in total? (\textbf{Hint:} DEL is the
  info field for a deletion. The -c option of the grep command can be
  used to return a count of matches.)
\item
  What type of event is predicted at IV:437148? What is the length of
  the SV? How many paired-end reads and split-reads support this SV
  variant call?
\item
  What is the total number of SV calls predicted on the IV chromosome?
\end{enumerate}

    Now continue to the next section of the tutorial:
\href{sv-calling.ipynb}{Calling structural variants}


    % Add a bibliography block to the postdoc



\newpage





    \hypertarget{calling-structural-variants}{%
\section{Calling Structural
Variants}\label{calling-structural-variants}}

There are several software tools available for calling structural
variants. We will use two callers in this part of the tutorial,
\texttt{breakdancer} and \texttt{lumpy}

\hypertarget{breakdancer}{%
\subsection{Breakdancer}\label{breakdancer}}

BreakDancer predicts five types of structural variants: insertions
(INS), deletions (DEL), inversions (INV), inter-chromosomal
translocations (CTX) and intra-chromosomal translocations (ITX) from
next-generation short paired-end sequencing reads using read pairs that
are mapped with unexpected separation distances or orientation.

Navigate to the \texttt{exercise2} directory:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{n+nb}{cd} ../exercise2
\end{Verbatim}
\end{tcolorbox}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ls
\end{Verbatim}
\end{tcolorbox}

    We will use the Breakdancer software package to call structural variants
on a yeast sample that was paired-end sequenced on the illumina HiSeq
2000. Breakdancer first needs to examine the BAM file to get information
on the fragment size distribution for each sequencing library contained
in the BAM file.

The breakdancer.config file has information about the sequencing library
fragment size distribution. Use the \texttt{cat} command to print the
contents of the \texttt{breakdancer.config} file.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    \textbf{Q What is the mean and standard deviation of the fragment size?}

Run the breakdancer SV caller using the command:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
breakdancer\PYZhy{}max breakdancer.config \PYZgt{} ERR1015121.breakdancer.out
\end{Verbatim}
\end{tcolorbox}

    Look at the output of Breakdancer.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
head ERR1015121.breakdancer.out
\end{Verbatim}
\end{tcolorbox}

    Note that the output from Breakdancer is NOT VCF format, instead it is a
simple text format with one line per SV event.

\hypertarget{exercises}{%
\subsubsection{Exercises}\label{exercises}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\item
  What type of SV event is predicted at position III:83065?
\item
  What is the size of this SV?
\item
  What is the score of this SV?
\item
  What type of SV event is predicted at position II:258766?
\item
  Convert the output of breakdancer into BED format
\end{enumerate}

    The BED format is explained here:
\url{https://genome.ucsc.edu/FAQ/FAQformat.html\#format1}

To complete this task, create a command that: 1. Extracts all the
deletions from the breakdancer.out file (\textbf{Hint:} use grep) 2.
Prints columns: 1, 2, 5, 7, and 9 to create a BED file with columns:
chromosome, start, end, name, and score. (\textbf{Hint:} use awk to do
this,
e.g.~\texttt{awk\ \textquotesingle{}\{print\ \$1"\textbackslash{}t"\$2\}\textquotesingle{}})
3. Print the resulting bed output into a file called:
breakdancer.dels.bed

    \hypertarget{inspecting-svs-with-igv}{%
\subsection{Inspecting SVs with IGV}\label{inspecting-svs-with-igv}}

Now we will open the IGV genome browser and inspect some of the
predicted structural variants.

To do this type:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
igv
\end{Verbatim}
\end{tcolorbox}

    Open the reference genome. \textbf{Go to ' \textit{Genomes -\textgreater{}
Load Genome From Server\ldots{}} ' and select ``S. cerevisiae
(SacCer3)''.}

Load the BAM file. \textbf{Go to ' \textit{File -\textgreater{} Load from
File\ldots{}} `. Select the ``ERR1015121.bam'' BAM file and click'
\textit{Open} '.}

Load the BED file for the deletion calls that you created in the
exercise 5 above. \textbf{Go to ' \textit{File -\textgreater{} Load from
File\ldots{}} `. Select the ``breakdancer.dels.bed'' BED file and click'
\textit{Open} '.}

\hypertarget{exercises}{%
\subsubsection{Exercises}\label{exercises}}

Using the navigation bar, go to region II:258,500-266,700.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\item
  Can you see the structural variant? What type of structural variant is
  it? (\textbf{Hint:} you may need to zoom out a little to see the full
  structural variant).
\item
  Can you see any evidence to support this SV call?
\item
  Can you estimate the size of the SV?
\end{enumerate}

The VCF in the \texttt{exercise1} directory was produced by another
structural variant caller on the same sample as this exercise.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{3}
\item
  Load the \texttt{exercise1/ERR1015121.vcf} VCF into IGV also (File -
  Load from file, and select \texttt{ERR1015121.vcf} in the exercise 1
  directory).
\item
  Was the structural variant at II:258766 also called by the other
  structural variant software (lumpy)? If so, what was the predicted
  size?
\end{enumerate}

Using the navigation bar, go to to region II:508,064-511,840.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{5}
\item
  Is there a SV deletion called in this region by either SV caller?
\item
  Is there any read support for a SV deletion in this region? If so, how
  many read pairs could support the deletion call (\textbf{Hint:} change
  the IGV view to \texttt{squished} and \texttt{View\ as\ pairs} to see
  any inconsistently aligned read pairs).
\end{enumerate}

    \hypertarget{lumpy}{%
\subsection{Lumpy}\label{lumpy}}

We will use the Lumpy software package to call structural variants on a
yeast sample that was paired-end sequenced on the Illumina Hiseq 2000.
Lumpy is designed to take BAM files that have been aligned with
\texttt{BWA-mem}.

Navigate to the \texttt{exercise\ 3} directory:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{n+nb}{cd} ../exercise3
\end{Verbatim}
\end{tcolorbox}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ls
\end{Verbatim}
\end{tcolorbox}

    Check that there is a BAM file called \texttt{ERR1015069.bam} and an
index file \texttt{ERR1015069.bam.bai} in the directory. The sequence
data has already been mapped with \texttt{bwa\ mem} and the results are
stored in \texttt{ERR1015069.bam}.

The first step for running Lumpy is to extract the read pairs that are
discordantly mapped (i.e.~pairs that are not mapped within the expected
fragment size distribution). Use Samtools to extract these reads:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
samtools view \PYZhy{}bh \PYZhy{}F \PY{l+m}{1294} ERR1015069.bam \PY{p}{|} samtools sort \PYZhy{}O bam \PYZhy{}T ERR1015069.temp \PYZhy{}o ERR1015069.discordants.bam
\end{Verbatim}
\end{tcolorbox}

    Now index the bam file. \textbf{Hint:} use samtools index

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    \hypertarget{exercises}{%
\subsubsection{Exercises}\label{exercises}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\item
  What does the \texttt{-F} option in \texttt{samtools\ view} do?
\item
  Which BAM flags does 1294 indicate? (\textbf{Hint:} in your web
  browser, visit
  https://broadinstitute.github.io/picard/explain-flags.html and enter
  1294 to find out)
\end{enumerate}

    The next step for Lumpy is to extract the reads that are only split
mapped (i.e.~split read alignments). This is all one single command:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
samtools view \PYZhy{}h ERR1015069.bam \PY{p}{|} extractSplitReads\PYZus{}BwaMem \PYZhy{}i stdin \PY{p}{|} samtools view \PYZhy{}b \PYZhy{} \PY{p}{|} samtools sort \PYZhy{}O bam \PYZhy{}T ERR1015069.temp \PYZhy{}o ERR1015069.splitters.bam
\end{Verbatim}
\end{tcolorbox}

    Now index the bam file. (\textbf{Hint:} use samtools index)

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    Finally, call structural variants using lumpy, providing it with the
original BAM file and the two BAM files we prepared earlier.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
lumpyexpress \PYZhy{}B ERR1015069.bam \PYZhy{}S ERR1015069.splitters.bam \PYZhy{}D ERR1015069.discordants.bam \PYZhy{}o ERR1015069.vcf
\end{Verbatim}
\end{tcolorbox}

    \hypertarget{exercises}{%
\subsubsection{Exercises}\label{exercises}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{2}
\item
  What type of SV event occurs at position IV:383393? What is the length
  of the SV event?
\item
  What type of SV event occurs at position IV:437149? What is the length
  of the SV event?
\end{enumerate}

    Congratulations, you have sucessfully called structural variants from
some NGS data. Now continue to the next section of the tutorial:
\href{long-reads.ipynb}{Calling structural variants from long reads}


    % Add a bibliography block to the postdoc



\newpage





    \hypertarget{calling-structural-variants-from-long-reads}{%
\section{Calling Structural Variants from Long
Reads}\label{calling-structural-variants-from-long-reads}}

    In this part of the tutorial we will use long read data to identify
structural variants using the SV caller Sniffles.

First navigate to the \texttt{exercise4} directory:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{n+nb}{cd} ../exercise4
\end{Verbatim}
\end{tcolorbox}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ls
\end{Verbatim}
\end{tcolorbox}

    \hypertarget{introducing-the-dataset}{%
\subsection{Introducing the dataset}\label{introducing-the-dataset}}

We will use data from a Saccharomyces cerevisiae strain (YPS128) that
was sequenced at the Wellcome Sanger Institute and deposited in the ENA
(Project: PRJEB7245, sample: SAMEA2757770, analysis: ERZ448241).

The sequencing reads are contained in a fastq file:

\texttt{YPS128.filtered\_subreads.10x.fastq.gz}

The reference genome is in the \texttt{../ref} directory in a fasta
file:

\texttt{Saccharomyces\_cerevisiae.R64-1-1.dna.toplevel.fa.gz}

    \hypertarget{align-the-data}{%
\subsection{Align the data}\label{align-the-data}}

Before you can use Sniffles to call SVs, it is very important that the
reads are aligned with an aligner suitable for long reads.

The software minimap2 is a long-read aligner designed to align PacBio or
Oxford Nanopore (standard and ultra-long) to a reference genome.

You can find the usage of minimap2 by typing:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
minimap2
\end{Verbatim}
\end{tcolorbox}

    To align the reads with minimap2, first index the reference for minimap2

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{n+nb}{cd} ../ref
minimap2 \PYZhy{}d Saccharomyces\PYZus{}cerevisiae.R64\PYZhy{}1\PYZhy{}1.dna.toplevel.fa.mmi Saccharomyces\PYZus{}cerevisiae.R64\PYZhy{}1\PYZhy{}1.dna.toplevel.fa.gz
\end{Verbatim}
\end{tcolorbox}

    Then use minimap2 to align the reads and send the output to a SAM file
called YPS128.filtered\_subreads.10x.sam

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    \textbf{Note:} use the -t option to use multiple threads in parallel
(this will increase the speed of the alignment by using more than one
CPU core, I suggest using 2). Also look at the -x option.

    Convert the SAM file to BAM.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    Sort the BAM file and produce a sorted BAM file called
\texttt{YPS128.filtered\_subreads.10X.sorted.bam}. \textbf{Hint:} Use
samtools sort.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    Use samtools calmd to calculates MD and NM tags. This enables variant
calling without requiring access to the entire original reference.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
gunzip ../ref/Saccharomyces\PYZus{}cerevisiae.R64\PYZhy{}1\PYZhy{}1.dna.toplevel.fa.gz
samtools calmd YPS128.filtered\PYZus{}subreads.10x.sorted.bam ../ref/Saccharomyces\PYZus{}cerevisiae.R64\PYZhy{}1\PYZhy{}1.dna.toplevel.fa \PY{p}{|}samtools view \PYZhy{}o YPS128.filtered\PYZus{}subreads.10x.fastq.sorted.calmd.bam \PYZhy{}O bam \PYZhy{}
\end{Verbatim}
\end{tcolorbox}

    Finally, use samtools to index this BAM file. \textbf{Hint} Use samtools
index

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    \hypertarget{call-structural-variants}{%
\subsection{Call structural variants}\label{call-structural-variants}}

Sniffles is a structural variation (SV) caller that is designed for long
reads (Pacbio or Oxford Nanopore). It detects all types of SVs (10bp+)
using evidence from split-read alignments, high-mismatch regions, and
coverage analysis. Sniffles takes the BAM file as input and outputs VCF.

To find the usage for Sniffles, type:

    sniffles

    Using the default parameters, call SVs with Sniffles and output the
results to a VCF file called \texttt{YPS128.10x.vcf}.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]

\end{Verbatim}
\end{tcolorbox}

    \textbf{Hint:} You don't need to change any of the default parameters,
but you will need to work out how to provide the input BAM file and
specify the output VCF file. The documentation on sniffles is here :
\url{https://github.com/fritzsedlazeck/Sniffles/wiki/Parameter}.

    \hypertarget{inspecting-svs-with-igv}{%
\subsection{Inspecting SVs with IGV}\label{inspecting-svs-with-igv}}

Open IGV:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
igv
\end{Verbatim}
\end{tcolorbox}

    Open the reference genome
\texttt{ref/Saccharomyces\_cerevisiae.R64-1-1.dna.toplevel.fa} and load
the BAM file
\texttt{YPS128.filtered\_subreads.10x.fastq.sorted.calmd.bam} and the
VCF file \texttt{YPS128.filtered\_subreads.10x.vcf}. Now answer the
questions that follow.

    \hypertarget{exercises}{%
\subsection{Exercises}\label{exercises}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  What sort of SV was called at on chromosome `XII' at position 451419?
\item
  What is the length of the SV?
\item
  How many reads are supporting the SV?
\item
  From a visual inspection of the SV in IGV, can you determine how
  accurate is the breakpoint of the called SV compared to what you see
  in IGV?
\end{enumerate}

    Now continue to the next section of the tutorial:
\href{bedtools.ipynb}{bedtools}


    % Add a bibliography block to the postdoc



\newpage





    \hypertarget{bedtools}{%
\section{Bedtools}\label{bedtools}}

Bedtools is an extremely useful tool for doing regional comparisons over
genomic co-ordinates. It has many commands for doing region based
comparisons with BAM, VCF, GFF, BED file formats.

To see the list of commands available, on the command line type:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
bedtools
\end{Verbatim}
\end{tcolorbox}

    Navigate to the \texttt{exercise5} directory.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{n+nb}{cd} ../exercise5
\end{Verbatim}
\end{tcolorbox}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ls
\end{Verbatim}
\end{tcolorbox}

    In this directory, there are two VCF files and the yeast genome
annotation in GFF3 format
\texttt{Saccharomyces\_cerevisiae.R64-1-1.82.genes.gff3}.

\hypertarget{bedtools-intersect}{%
\subsection{bedtools intersect}\label{bedtools-intersect}}

Given two sets of genomic features, the \texttt{bedtools\ intersect}
command can be used to determine whether or not any of the features in
the two sets ``overlap'' with one another. For the intersect command,
the -a and -b parameters are used to denote the input files A and B.

For example, to find out the overlap between the SVs in
\texttt{ERR1015069.dels.vcf} and the annotated region of the yeast
genome try

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
bedtools intersect \PYZhy{}a ERR1015069.dels.vcf \PYZhy{}b Saccharomyces\PYZus{}cerevisiae.R64\PYZhy{}1\PYZhy{}1.82.genes.gff3
\end{Verbatim}
\end{tcolorbox}

    This command reports the variant in the file
\texttt{ERR1015069.dels.vcf} every time it overlaps with a feature in
\texttt{Saccharomyces\_cerevisiae.R64-1-1.82.genes.gff3}. Therefore if a
variant overlaps more than one feature it will be reported more than
once. To report the unique set of variants use:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
bedtools intersect \PYZhy{}u \PYZhy{}a ERR1015069.dels.vcf \PYZhy{}b Saccharomyces\PYZus{}cerevisiae.R64\PYZhy{}1\PYZhy{}1.82.genes.gff3
\end{Verbatim}
\end{tcolorbox}

    The default is to report overlaps between features in A and B so long as
at least one base pair of overlap exists. However, the \texttt{-f}
option allows you to specify what fraction of each feature in A should
be overlapped by a feature in B before it is reported.

To specify a more strict intersect and require at least 25\% of overlap
of the SV with the genes use the command:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
bedtools intersect \PYZhy{}u \PYZhy{}f \PY{l+m}{0}.25 \PYZhy{}a ERR1015069.dels.vcf \PYZhy{}b Saccharomyces\PYZus{}cerevisiae.R64\PYZhy{}1\PYZhy{}1.82.genes.gff3
\end{Verbatim}
\end{tcolorbox}

    The \texttt{bedtools\ intersect} command can also be used to determine
how many SVs overlap between two VCF files. For more information about
\texttt{bedtools\ intersect} see the help:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
bedtools intersect \PYZhy{}h
\end{Verbatim}
\end{tcolorbox}

    \hypertarget{bedtools-closest}{%
\subsection{bedtools closest}\label{bedtools-closest}}

Similar to intersect, \texttt{bedtools\ closest} searches for
overlapping features in A and B. In the event that no feature in B
overlaps the current feature in A, closest will report the nearest (that
is, least genomic distance from the start or end of A) feature in B.

An example of the usage of \texttt{bedtools\ closest} is:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
bedtools closest \PYZhy{}d \PYZhy{}a ERR1015069.dels.vcf \PYZhy{}b Saccharomyces\PYZus{}cerevisiae.R64\PYZhy{}1\PYZhy{}1.82.genes.gff3
\end{Verbatim}
\end{tcolorbox}

    This command will list all the features in the file
\texttt{Saccharomyces\_cerevisiae.R64-1-1.82.genes.gff3} that are
closest to each of the variants in \texttt{ERR1015069.dels.vcf}.

The \texttt{-d} option means that in addition to the closest feature in
\texttt{Saccharomyces\_cerevisiae.R64-1-1.82.genes.gff3}, the distance
to the variant in \texttt{ERR1015069.dels.vcf} will be reported as an
extra column. The reported distance for any overlapping features will be
0.

For example, to find the closest gene to the variant found at position
43018 on chromosome XV, try

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
bedtools closest \PYZhy{}d \PYZhy{}a ERR1015069.dels.vcf \PYZhy{}b Saccharomyces\PYZus{}cerevisiae.R64\PYZhy{}1\PYZhy{}1.82.genes.gff3\PY{p}{|} grep XV \PY{p}{|} grep \PY{l+m}{43018}
\end{Verbatim}
\end{tcolorbox}

    For more information about \texttt{bedtools\ closest} see the help:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
bedtools closest \PYZhy{}h
\end{Verbatim}
\end{tcolorbox}

    \hypertarget{exercises}{%
\subsection{Exercises}\label{exercises}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\item
  How many SVs found in \texttt{ERR1015069.dels.vcf} overlap with a
  gene? (\textbf{Hint:} Use bedtools intersect command)
\item
  How many SVs found in \texttt{ERR1015069.dels.vcf} do not overlap with
  a gene? (\textbf{Hint:} note the -v parameter to bedtools intersect)
\item
  How many SVs found in \texttt{ERR1015069.dels.vcf} overlap with a more
  strict definition of 50\%?
\item
  What is the closest gene to the structural variant at IV:384220 in
  \texttt{ERR1015069.dels.vcf}?
\item
  How many SVs overlap between the two files
  \texttt{ERR1015069.dels.vcf} and \texttt{ERR1015121.dels.vcf}?
\item
  How many SVs have a 50\% reciprocal overlap between the two files
  \texttt{ERR1015069.dels.vcf} and \texttt{ERR1015121.dels.vcf}
  (\textbf{Hint:} first find the option for reciprocal overlap by
  typing: bedtools intersect -h)
\end{enumerate}

    Congratulations, you have reached the end of the tutorial.


    % Add a bibliography block to the postdoc



\end{document}
